avail/src/main/kotlin/avail/anvil/Styles.kt

```diff
@@ -3559,6 +3559,32 @@ object RenderingEngine
 				}
 			}
 		}
+		// Add the information about unused declarations.
```

Comment:
I'm guessing the goal here is that this would allow highlighting in the IDE unused variables to make the user aware that those can be safely removed. If that's the case, are we supporting this in other types of declarations like methods or classes or just variables?

Reply:
Yes, this is to make unused declarations easier to capture and (with some more work) style.  Currently, this only applies to local things, which can be function arguments, local constants, local variables, primitive failure variables, and block labels, whether referred to directly in the scope in which they're defined, or inside nested scopes, where the captured variables or values are captured in "outer" variables of the inner function.  The `L1_doClose` instruction accomplishes that last bit.
Module compilation passes through lexing and parsing into phrases (see `PhraseDescriptor` hierarchy), including type/grammar checking and macro expansion, then the outer statement (`A_Phrase`) gets wrapped as the sole statement of a zero-argument block phrase, which gets compiled via `AvailCodeGenerator` into a sequence of `AvailInstruction`s, where last-use analysis happens, and not much else, prior to generating the actual `A_RawFunction` (which has a `CompiledCodeDescriptor`), which contains the nybblecode instructions and literals.  At some point where the function is to be pushed on the stack, the raw function (aka compiled code) is combined with any outer variables that it needs via the `L1_doClose` instruction, which constructs an `A_Function` (FunctionDescriptor), which then gets pushed.  If the raw function doesn't need to capture any locals, the function itself is constructed and stored as literal, which simply gets pushed instead of using `L1_doClose`.
Getting back to your question, method usages are tracked in a different way, via a compiled module's `NamesIndex`, which keeps a somewhat complicated index into the `PhrasePathRecord`, which captures a forest of simplified phrases for that module compilation, along with definitions and uses of each name (`A_Atom`), and the location of each token.
Types are not separately indexed at the moment, as declaration of a type generally involves the automatic construction of a zero-argument method that returns the type as a constant (typically at compile time).  For example, "`map`" is a zero-argument method that returns the most general map type.  That keeps the decision about what to publish uniform, as only atoms (the names of methods, and rarely lexers or other things) may be exported and imported between modules (module constants and variables cannot be exported as such).  If you can see a name in your module, you can use it.  This differs from C++'s mechanism where you can see "everything", but only use things that have the right public/protected/friend permission.  The mechanism of visibility-is-permission is an essential component of large scale modularity.

Reply:
Thank you for reviewing the phases and how they work; this helps! Is there a document or someplace where I can look at the Avail VM instructions? I'll go over the types and sources you mentioned here to understand how the Compile and runtime of Avail work.

Reply:
Usually there's a short class hierarchy or enum class for things like interpreter instructions or Avail type hierarchies.  I'm actually currently tasked with filling in some of those blanks as formal /avail/docs READMEs.  There are lots of scattered whitepapers and pictures of whiteboards from over the years, covering a range from still correct to entirely defunct.  I'll tag you to review them as I commit them to /avail/docs.

Reply:
Specifically, the L1 instructions, also called nybblecodes, are the enum values in `L1Operation`.  The enum values with ordinals 0-14 are written with a single nybble (plus operands), and the remaining operations write the extension nybble 15 followed by another nybble.  Operand encoding is handled by AvailInstruction.kt:158, writeIntegerOn().  Values 0-9 take a nybble, 10-57 take a full byte, 58-313 take 1.5 bytes, etc).
These are decoded at runtime by `L1InstructionDecoder`, found in `CompiledCodeDescriptor.kt:373`, specifically `getOperand()` on :478.  The constant on :482 is a bit less confusing after you negate it, or convert it to a ULong.  It's written to avoid branches as much as possible (other than the counter-based loop), without introducing deep arithmetic pipeline dependence.
Note that functions that consume any significant fraction of the execution time get translated quickly into an `L2SimpleChunk`, which eliminates the nybblecode decoding cost.  If that new form continues to consume significant time, it's translated through a slower mechanism into an `L2JVMChunk`, which involves dynamic Java bytecode generation and subsequent class loading from a byte array.  This last translation is where we do flow analysis, inlined method lookups, primitive-specific code generation, common subexpression elimination, register coloring, and a bunch of other stuff.  It does not yet support code splitting or inlining of called non-primitive methods – I have early partial successes on a stale branch from a year ago.  We expect another significant performance jump for warmed-up code once that's implemented.  Anvil, styling, debugging, and indexing have been a much higher priority for some time.

Reply:
There are other interesting mini-languages besides the L1 nybblecodes.  Parsing, for example, is mostly done via the `ParsingOperation` sealed hierarchy (it was an enum until Todd's recent refactoring).  Each visible method or macro definition produces a parsing plan (`A_DefinitionParsingPlan` / `DefinitionParsingPlanDescriptor`), all of which get added to the root `A_BundleTree` (`MessageBundleTreeDescriptor`) for the current module.  Parsing an invocation of a method or macro involves traversing that bundle tree along every applicable path, performing it on the current parsing context.  Some actions just rearrange the current stack of expressions, like popping the last N parsed phrases and pushing a list phrase of length N.  Some operations jump to a different position in the operation list.  A branch operation continues at the next instruction AND jumps forward to another position, by queueing actions in a thread worker pool.  Another operation checks if the last pushed phrase would yield a value of a particular type.  Another operation runs a macro prefix function, supplied at definition time to correspond with the location of a "`§`" in the macro name.  Another consumes a particular token string (giving up that parse theory if that token isn't next in the source).  Another operation effectively reenters the parser to parse an argument expression where a "`_`" occurs in the method name.
The bundle tree groups these parsing plans by their next instruction, so that if 10 parsing plans want to swap the top two elements of the expression stack, it only has to do it once.  Or if 10 parsing plans are looking for a "`[`" token next, it only has to do the test once.  In fact, a bundle tree holds a map of token string -> next bundle tree, so it just has to see what's next and look it up in the map, ignoring all the other map entries.
There's an additional mechanism in there called an `AvailCompilerBipartiteRendezvous`, which contains a list of actions and a list of solutions.  When a parsing plan hits a point where it wants to parse a subexpression (to fill a "`_`"), it looks up the bipartite rendezvous that may have been created for the current parse position in the source, creating a new one if needed.  It then adds an action to that rendezvous, running it with any solutions (parsed subexpressions) that have been found at that position.  Symmetrically, when new solutions are found, they go into the rendezvous's list of solutions, and each saved action is executed with that new solution.  This ensures that every action runs with every solution, regardless of the order they arrive.  Only when the first action arrives, creating the rendezvous, does it launch tasks to produce all the solutions at that position.  Subsequent actions just assume those tasks have already been started, and will eventually invoke the action with each solution.  I think this probably needs some examples, and /avail/docs is the right place to write it up.

Another area of potential interest is subclasses of `Primitive`, which all start with `P_`.  For example, `P_Addition` is how numbers get added.  In `Fallible Primitives.avail:2865`, the `"_+_"` operation is defined, and on :2870 it says `Primitive Addition`..., which indicates the `P_Addition` primitive should be invoked (it prepends the "`P_`").  When running 1 + 2, it calls `"_+_"` with those arguments, which eventually end up in `P_Addition.kt:84`, its `attempt()` method.  That sends the AvailObject representing 1 the message `plusCanDestroy()`, which itself sends the message `o_PlusCanDestroy()` to the `IntegerDescriptor` that the 1 object was using.  It also would have worked for any numeric type, not just an integer, since `o_PlusCanDestroy()` has definitions in all the subclasses of `AbstractNumberDescriptor`, not just `IntegerDescriptor`.  A double-dispatch mechanism ensures all kinds of numbers can be correctly added to all other kinds of numbers (but nothing else).  Now I think you see why we sometimes call those `o_*()` operations sub-primitives.